configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java

```diff
@@ -252,17 +253,77 @@ private Deployment deployFromPreparedSession(LocalSession session, Tenant tenant
     // ---------------- Application operations ----------------------------------------------------------------
 
     /**
-     * Removes a previously deployed application
+     * Deletes an application
      *
-     * @return true if the application was found and removed, false if it was not present
-     * @throws RuntimeException if the remove transaction fails. This method is exception safe.
+     * @return true if the application was found and deleted, false if it was not present
+     * @throws RuntimeException if the delete transaction fails. This method is exception safe.
      */
-    public boolean remove(ApplicationId applicationId) {
+    public boolean delete(ApplicationId applicationId) {
+        // TODO: Use deleteApplication() in all zones, for now use it only in non-hosted
+        if (configserverConfig.hostedVespa()) {
+            return deleteApplicationLegacy(applicationId);
+        } else {
+            return deleteApplication(applicationId);
+        }
+    }
+
+    /**
+     * Deletes an application
+     *
+     * @return true if the application was found and deleted, false if it was not present
+     * @throws RuntimeException if the delete transaction fails. This method is exception safe.
+     */
+    public boolean deleteApplication(ApplicationId applicationId) {
+        Tenant tenant = tenantRepository.getTenant(applicationId.tenant());
+        if (tenant == null) return false;
+
+        TenantApplications tenantApplications = tenant.getApplicationRepo();
+        if (!tenantApplications.listApplications().contains(applicationId)) return false;
+
+        // Deleting an application is done by deleting the remote session and waiting
+        // until the config server where the deployment happened picks it up and deletes
+        // the local session
+        long sessionId = tenantApplications.getSessionIdForApplication(applicationId);
+        RemoteSession remoteSession = getRemoteSession(tenant, sessionId);
+        remoteSession.createDeleteTransaction().commit();
+
```

Comment:
How do we handle the state we'll be in if this line (288) succeeds but the transaction created below fails?

Perhaps you could move the second transaction to LocalSession.deleteSession such that all of this could be in one transaction (per node)?

Reply:
Yeah, this should all be done in one transaction. As it is today we don't have the information needed to do all the operations when all we have is the session and we want to delete it. There might be some other way to do it, though, I'll add a TODO in a later PR and look into it.